### PR TITLE
Implemented bitwise-instructions

### PR DESCRIPTION
**Pull request by BarbaraButz**
~~Please merge #23 first.~~
Also check the mux[i] and sadd[i] instructions especially careful for mistakes.

### CODE REVIEW

**Comment by BarbaraButz** [comment]
~~I didn't manage to update the branch without merging/ with only rebasing. Pls help.~~
Branch can be safely merged now.

File: src/machine/behavior/is/bitwise/and.rs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 use machine::state::State;
 
+/// bitwise and
```
Comment (teiesti) [inline src/machine/behavior/is/bitwise/and.rs:3]: I think, the doc-comments should be a little bit more detailed. But I suggest to do this in the end, when we have finished the other task.

File: src/machine/behavior/is/bitwise/mux.rs [hunk 2]
```diff
@@ -1,6 +1,15 @@
...
-pub fn mux(_state: &mut State, _x: u8, _y: u8, _z: u8) {
-    unimplemented!();
-}
+/// bitwise multiplex
+pub fn mux(state: &mut State, x: u8, y: u8, z: u8) {
+    // Load operands
+    let op1: u64 = state.gpr[y].into();
+    let op2: u64 = state.gpr[z].into();
+    let op3: u64 = state.sr[R::M].into();
+    // Execute
```
Comment (teiesti) [inline src/machine/behavior/is/bitwise/mux.rs]: There is a newline missing.

File: src/machine/behavior/is/bitwise/sadd.rs [hunk 3]
```diff
@@ -1,6 +1,18 @@
 use machine::state::State;
 
-pub fn sadd(_state: &mut State, _x: u8, _y: u8, _z: u8) {
-    unimplemented!();
+/// sideways add
+pub fn sadd(state: &mut State, x: u8, y: u8, z: u8) {
+    // calculate andn and load result
+    super::andn(state, x, y, z);
```
Comment (teiesti) [inline src/machine/behavior/is/bitwise/sadd.rs]: I am not sure if it is a good idea to call another instruction from here!

**Review by teiesti — CHANGES_REQUESTED** [review]

File: src/machine/behavior/is/bitwise/sadd.rs (hunk 3 above)
Comment (KarolinePlum) [reply src/machine/behavior/is/bitwise/sadd.rs]: Why not?

Comment (teiesti) [reply src/machine/behavior/is/bitwise/sadd.rs]: What if we want to implement instruction tracing within these functions?

Comment (KarolinePlum) [reply src/machine/behavior/is/bitwise/sadd.rs]: Wouldn't that mean lots of redundant code?
Why not simply trace which function is called by the control unit? I'm not sure that I've understood correctly what you mean with "instruction tracing".

**Comment by BarbaraButz** [comment]
~~I will implement the requested changes (except for documentation) as soon as the function-call-question is resolved.~~

File: src/machine/behavior/is/bitwise/sadd.rs (hunk 3 above)
Comment (teiesti) [reply src/machine/behavior/is/bitwise/sadd.rs]: @KarolinePlum is right: You can call instructions from other instructions.

File: src/machine/behavior/is/bitwise/saddi.rs [hunk 4]
```diff
@@ -1,6 +1,20 @@
 use machine::state::State;
 
-pub fn saddi(_state: &mut State, _x: u8, _y: u8, _z: u8) {
-    unimplemented!();
-}
+/// sideways add immediate
+pub fn saddi(state: &mut State, x: u8, y: u8, z: u8) {
+    // calculate andn and load result
```
Comment (teiesti) [inline src/machine/behavior/is/bitwise/saddi.rs]: Woud be nice, if this starts with a capital letter.

**Review by teiesti — CHANGES_REQUESTED** [review]

File: src/machine/behavior/is/bitwise/sadd.rs [hunk 5]
```diff
@@ -1,6 +1,20 @@
...
-}
+/// sideways add
+pub fn sadd(state: &mut State, x: u8, y: u8, z: u8) {
+    // Calculate andn and load result
+    super::andn(state, x, y, z);
+    let mut bits: u64 = state.gpr[x].into();
+
+    // Execute
+    let mut res: u64 = 0;
+    while bits != 0 {
```
Comment (teiesti) [inline src/machine/behavior/is/bitwise/sadd.rs:11]: I am pretty sure this implementation is not right.

File: src/machine/behavior/is/bitwise/saddi.rs [hunk 6]
```diff
@@ -1,6 +1,20 @@
...
-}
+/// sideways add immediate
+pub fn saddi(state: &mut State, x: u8, y: u8, z: u8) {
+    // Calculate andn and load result
+    super::andni(state, x, y, z);
+    let mut bits: u64 = state.gpr[x].into();
+
+    // Execute
+    let mut res: u64 = 0;
+    while bits != 0 {
```
Comment (teiesti) [inline src/machine/behavior/is/bitwise/saddi.rs:11]: Same as SADD.

**Review by teiesti — CHANGES_REQUESTED** [review]

File: src/machine/behavior/is/bitwise/sadd.rs (hunk 5 above)
Comment (BarbaraButz) [reply src/machine/behavior/is/bitwise/sadd.rs:11]: Not wrong, but unnecessary complicated. Will prettify.

**Review by teiesti — APPROVED** [review]

**Pull request merged**